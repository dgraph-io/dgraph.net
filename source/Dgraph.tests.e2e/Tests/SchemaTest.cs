using System.Threading.Tasks;
using Assent;
using Dgraph.tests.e2e.Orchestration;
using FluentAssertions;
using Newtonsoft.Json;
using Dgraph.Schema;
using Api;
using System;
using System.Threading;

namespace Dgraph.tests.e2e.Tests
{
    public class SchemaTest : DgraphDotNetE2ETest {
        public SchemaTest(DgraphClientFactory clientFactory) : base(clientFactory) { }

        public async override Task Test() {
            using(var client = await ClientFactory.GetDgraphClient()) {
                await InitialSchemaIsAsExpected(client);
                await AlterSchemAsExpected(client);
                await AlterSchemaAgainAsExpected(client);
                await SchemaQueryWithRestrictions(client);

                await ErrorsResultInFailedQuery(client);
            }
        }

        private async Task InitialSchemaIsAsExpected(IDgraphClient client) {
            var response = await client.NewReadOnlyTransaction().Query("schema {}");
            AssertResultIsSuccess(response);

            var schema = JsonConvert.DeserializeObject<DgraphSchema>(response.Value.Json);
            this.Assent(schema.ToString(), AssentConfiguration);
        }

        private async Task AlterSchemAsExpected(IDgraphClient client) {
            var alterSchemaResult = await client.Alter(
                new Operation{ Schema = ReadEmbeddedFile("test.schema") });
            AssertResultIsSuccess(alterSchemaResult);

            // After an Alter, Dgraph computes indexes in the background.
            // So a first schema query after Alter might return a schema
            // without indexes.  We could poll and backoff and show that
            // ... but we aren't testing that here, just that the schema
            // updates.
            Thread.Sleep(TimeSpan.FromSeconds(5));

            var response = await client.NewReadOnlyTransaction().Query("schema {}");
            AssertResultIsSuccess(response);

            var schema = JsonConvert.DeserializeObject<DgraphSchema>(response.Value.Json);
            this.Assent(schema.ToString(), AssentConfiguration);
        }

        private async Task AlterSchemaAgainAsExpected(IDgraphClient client) {
            var alterSchemaResult = await client.Alter(
                new Operation{ Schema = ReadEmbeddedFile("altered.schema") });
            AssertResultIsSuccess(alterSchemaResult);

            Thread.Sleep(TimeSpan.FromSeconds(5));

            var response = await client.NewReadOnlyTransaction().Query("schema {}");
            AssertResultIsSuccess(response);

            var schema = JsonConvert.DeserializeObject<DgraphSchema>(response.Value.Json);
            this.Assent(schema.ToString(), AssentConfiguration);
        }

        private async Task SchemaQueryWithRestrictions(IDgraphClient client) {
            var response = await client.NewReadOnlyTransaction().Query(
                "schema(pred: [name, friends, dob, scores]) { type }");
            AssertResultIsSuccess(response);

            var schema = JsonConvert.DeserializeObject<DgraphSchema>(response.Value.Json);
            this.Assent(schema.ToString(), AssentConfiguration);
        }

        private async Task ErrorsResultInFailedQuery(IDgraphClient client) {
            // maformed
            var q1result = await client.NewReadOnlyTransaction().Query(
                "schema(pred: [name, friends, dob, scores]) { type ");
            q1result.IsSuccess.Should().BeFalse();
        }
    }
}