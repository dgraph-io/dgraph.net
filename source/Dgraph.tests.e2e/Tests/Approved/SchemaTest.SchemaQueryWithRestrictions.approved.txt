dob: datetime .
friends: uid .
name: string .
scores: int .
